File: 0131-palindrome-partitioning/0131-palindrome-partitioning.java
class Solution {
    List<List<String>> ans = new ArrayList<>();
    public boolean isPalindrome(String s, int l, int r) {
        while (l < r) {
            if (s.charAt(l++) != s.charAt(r--)) {
                return false;
            }
        }
        return true;
    }
    public void helper(int index, String s, List<String> curr) {
        int n = s.length();
        if (index == n) {
            ans.add(new ArrayList<>(curr));
            return;
        }
        for (int i = index; i < n; i++) {
            if (isPalindrome(s, index, i)) {
                curr.add(s.substring(index, i+1));
                helper(i + 1, s, curr);
                curr.remove(curr.size() - 1);
            }
        }
    }
    public List<List<String>> partition(String s) {
        helper(0, s, new ArrayList<>());
        return ans;
    }
}

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.cpp
class Solution {
public:
    vector<vector<string>> ans;
    int n;
    bool isPalindrome(string& s, int l, int r) {
        while (l < r) {
            if (s[l++] != s[r--]) return false;
        }
        return true;
    }
    void helper(int index, string& s, vector<string>& curr) {
        n = s.size();
        if (index == n) {
            ans.push_back(curr);
        }
        for (int i = index; i < n; i++) {
            if (isPalindrome(s, index, i)){
                curr.push_back(s.substr(index, i-index+1));
                helper(i+1, s, curr);
                curr.pop_back();
            }
        }
    }
    vector<vector<string>> partition(string s) {
        vector<string> curr;
        helper(0, s, curr);
        return ans;
    }
};